File: SampleProjects/GoT-Characters-master/Objective-C/GoT/Loader.m
//
//  Loader.m
//  GoT
//
//  Created by Paciej on 21/10/15.
//  Copyright © 2015 Maciej Piotrowski. All rights reserved.
//

#import "Loader.h"
#import "NSData+JSON.h"

@interface Loader ()

@property(nonatomic, strong, nonnull) NSString *webserviceURLString;
@property(nonatomic, strong, nonnull) NSURLSession *session;

@end

@implementation Loader

- (nonnull instancetype)
initWithWebserviceURLString:(nonnull NSString *)urlString
session:(nonnull NSURLSession *)session {
    self = [super init];
    if (self) {
        self.webserviceURLString = urlString;
        self.session = session;
    }
    return self;
}

- (void)loadAsynchronously:(nonnull AsyncLoadConfiguration *)configuration
                  callback:(nullable void (^)(id _Nullable result))callback {
    NSString *urlString =
    [NSString stringWithFormat:@"%@%@%@", self.webserviceURLString,
     configuration.webserviceEndpoint,
     configuration.webserviceQuery]; //TODO: use NSURLQueryItem  https://littlebitesofcocoa.com/128-nsurlqueryitem-nsurlcomponents
    NSURLComponents *components = [[NSURLComponents alloc] initWithString:urlString];
    NSURLSessionDataTask *task = [self.session dataTaskWithURL:components.URL
                completionHandler:^(NSData *_Nullable data,
                                    NSURLResponse *_Nullable response,
                                    NSError *_Nullable error) {
                    if (error) {
                        NSLog(@"loadAsynchronously error: %@",
                              error.debugDescription);
                        callback(nil);
                        return;
                    }
                    id parsedObject = configuration.responseParsingBlock(data);
                    callback(parsedObject);
                }];
    [task resume];
}

@end

File: SampleProjects/GoT-Characters-master/Swift/GoT/Loader.swift
//
//  Loader.swift
//  GoT
//
//  Created by Paciej on 21/10/15.
//  Copyright © 2015 Maciej Piotrowski. All rights reserved.
//

import Foundation

class Loader: NSObject {
    private var webserviceURLString = ""
    private var session: URLSession!

    init(webserviceURLString urlString: String, session: URLSession) {
        super.init()
        
        webserviceURLString = urlString
        self.session = session
    
    }

    func loadAsynchronously(_ configuration: AsyncLoadConfiguration, callback: ((_ result: Any?) -> Void)? = nil) {
        let urlString = "\(webserviceURLString)\(configuration.webserviceEndpoint)\(configuration.webserviceQuery ?? "")" //TODO: use NSURLQueryItem  https://littlebitesofcocoa.com/128-nsurlqueryitem-nsurlcomponents
        let components = URLComponents(string: urlString)
        var task: URLSessionDataTask? = nil
        if let anURL = components?.url {
            task = session.dataTask(with: anURL, completionHandler: { data, response, error in
                if error != nil {
                    print("loadAsynchronously error: \(error.debugDescription)")
                    callback?(nil)
                    return
                }
                let parsedObject = configuration.responseParsingBlock?(data!)
                callback?(parsedObject)
            })
        }
        task?.resume()
    }
}